bitcoin/src/address/mod.rs

```diff
@@ -802,18 +868,18 @@ impl<V: NetworkValidation> fmt::Debug for Address<V> {
 }
 
 /// Address can be parsed only with `NetworkUnchecked`.
-impl FromStr for Address<NetworkUnchecked> {
+impl<U: NetworkValidationUnchecked> FromStr for Address<U> {
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Address<NetworkUnchecked>, ParseError> {
+    fn from_str(s: &str) -> Result<Self, ParseError> {
         if let Ok((hrp, witness_version, data)) = bech32::segwit::decode(s) {
             let version = WitnessVersion::try_from(witness_version)?;
             let program = WitnessProgram::new(version, &data)
                 .expect("bech32 guarantees valid program length for witness");
 
             let hrp = KnownHrp::from_hrp(hrp)?;
             let inner = AddressInner::Segwit { program, hrp };
-            return Ok(Address(inner, PhantomData));
+            return Ok(Address(inner, PhantomData::<U>));
```

Comment:
How the hell can inference fail here?! (This is not a code bug, I'm just confused about rustc.)

Reply:
No clue, I was also surprised.